### PR TITLE
prometheus-nginx-exporter: switch to official version & general prometheus-exporters cleanup

### PR DESCRIPTION
**Pull request by WilliButz**
###### Motivation for this change
NGINX Inc. provide their own prometheus exporter.

###### Things done
- fixed two failing prometheus-exporter tests (bind and varnish)
- refactored the prometheus-exporter "meta module" a bit
- added functionality to define option renamings and removals in exporter submodules
- added a small documentation section on adding option overrides for exporter submodules
- switched to the official version of the nginx-exporter and updated the module
- replaced 'with lib;' in the exporter "meta module" and the exporter test, to avoid having implicit imports in these more complex files. 

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS (x86_64-linux & aarch64-linux)
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
/cc @fpletz  @benley as additional maintainers of the package `prometheus-nginx-exporter`

### CODE REVIEW

**Comment by WilliButz** [comment]
@GrahamcOfBorg test prometheus-exporters.bind prometheus-exporters.blackbox prometheus-exporters.collectd prometheus-exporters.dnsmasq prometheus-exporters.dovecot prometheus-exporters.fritzbox prometheus-exporters.json prometheus-exporters.nginx prometheus-exporters.node
prometheus-exporters.postfix prometheus-exporters.snmp prometheus-exporters.surfboard prometheus-exporters.tor prometheus-exporters.varnish prometheus-exporters.wireguard

File: nixos/tests/prometheus-exporters.nix [hunk 1]
```diff
@@ -73,7 +74,7 @@ let
       exporterTest = ''
         waitForUnit("prometheus-bind-exporter.service");
         waitForOpenPort(9119);
-        succeed("curl -sSf http://localhost:9119/metrics" | grep -q 'bind_query_recursions_total 0');
+        succeed("curl -sSf http://localhost:9119/metrics | grep -q 'bind_query_recursions_total 0'");
```
Comment (benley) [inline nixos/tests/prometheus-exporters.nix]: heh, I guess that test never worked...

Comment (WilliButz) [reply nixos/tests/prometheus-exporters.nix]: yep :)

File: pkgs/servers/monitoring/prometheus/nginx-exporter.nix [hunk 2]
```diff
@@ -2,23 +2,25 @@
...
+    sha256 = "0c5bxl9xrd4gh2w5wyrzghmbcy9k1khydzml5cm0rsyqhwsvs8m5";
   };
 
-  goDeps = ./nginx-exporter_deps.nix;
-
   meta = with stdenv.lib; {
-    description = "Metrics relay from nginx stats to Prometheus";
-    homepage = https://github.com/discordianfish/nginx_exporter;
-    license = licenses.mit;
+    description = " NGINX Prometheus Exporter for NGINX and NGINX Plus ";
```
Comment (benley) [inline pkgs/servers/monitoring/prometheus/nginx-exporter.nix]:
```suggestion
    description = "Prometheus Exporter for NGINX and NGINX Plus";
```

File: nixos/modules/services/monitoring/prometheus/exporters/varnish.nix [hunk 3]
```diff
@@ -68,7 +68,6 @@ in
   serviceOpts = {
     path = [ pkgs.varnish ];
     serviceConfig = {
-      DynamicUser = true;
```
Comment (benley) [inline nixos/modules/services/monitoring/prometheus/exporters/varnish.nix]: I'm curious, does DynamicUser not work here?

File: pkgs/servers/monitoring/prometheus/nginx-exporter.nix (hunk 2 above)
Comment (WilliButz) [reply pkgs/servers/monitoring/prometheus/nginx-exporter.nix]: Oh, I didn't notice the phrasing before. It's copy-pasta from the official repo's description: https://github.com/nginxinc/nginx-prometheus-exporter

File: nixos/modules/services/monitoring/prometheus/exporters/varnish.nix (hunk 3 above)
Comment (WilliButz) [reply nixos/modules/services/monitoring/prometheus/exporters/varnish.nix]: It needs to access /var/spool/varnish. Running as dynamic user, the exporter is unable to find varnish.

Comment (benley) [reply nixos/modules/services/monitoring/prometheus/exporters/varnish.nix]: Makes sense, thanks.

File: nixos/modules/services/monitoring/prometheus/exporters/nginx.nix [hunk 4]
```diff
@@ -16,32 +16,40 @@ in
...
+          --web.listen-address ${cfg.listenAddress}:${toString cfg.port} \
+          --web.telemetry-path ${cfg.telemetryPath} \
           ${concatStringsSep " \\\n  " cfg.extraFlags}
       '';
     };
   };
+  imports = [
+    (mkRenamedOptionModule [ "telemetryEndpoint" ] [ "telemetryPath" ])
+    (mkRemovedOptionModule [ "insecure" ] ''
+      This option was replaced by 'prometheus.exporters.nginx.sslVerify' which defaults to true.
```
Comment (benley) [inline nixos/modules/services/monitoring/prometheus/exporters/nginx.nix]: Restating the other option's default here is redundant (even though the default is flipped - the option names make the behaviour seem pretty obvious, imho)

```suggestion
      This option was replaced by 'prometheus.exporters.nginx.sslVerify' .
```

Comment (WilliButz) [reply nixos/modules/services/monitoring/prometheus/exporters/nginx.nix]: :+1:

File: pkgs/servers/monitoring/prometheus/nginx-exporter.nix (hunk 2 above)
Comment (benley) [reply pkgs/servers/monitoring/prometheus/nginx-exporter.nix]: Fair enough.  Maybe just trim the surrounding whitespace?

Comment (WilliButz) [reply pkgs/servers/monitoring/prometheus/nginx-exporter.nix]: Yes, sounds good :+1:

**Comment by risicle** [comment]
`nox-review` is happy enough on macos 10.13 and the `nginx-prometheus-exporter` binary runs.

**Comment by WilliButz** [comment]
I included the proposed changes and rebased onto master.
@globin reminded me to also add a small section in the 19.09 release notes about the incompatible metrics.

**Review by fpletz — APPROVED** [review]

**Pull request merged**